crates/hir_ty/src/infer/coerce.rs

```diff
@@ -5,30 +5,52 @@
 //! See <https://doc.rust-lang.org/nomicon/coercions.html> and
 //! `librustc_typeck/check/coercion.rs`.
 
-use chalk_ir::{cast::Cast, Mutability, TyVariableKind};
+use chalk_ir::{cast::Cast, Goal, Mutability, TyVariableKind};
 use hir_def::{expr::ExprId, lang_item::LangItemTarget};
 
 use crate::{
-    autoderef, infer::TypeMismatch, static_lifetime, Canonical, DomainGoal, FnPointer, FnSig,
-    Interner, Solution, Substitution, Ty, TyBuilder, TyExt, TyKind,
+    autoderef,
+    infer::{Adjust, Adjustment, AutoBorrow, PointerCast, TypeMismatch},
+    static_lifetime, Canonical, DomainGoal, FnPointer, FnSig, Interner, Solution, Substitution, Ty,
+    TyBuilder, TyExt, TyKind,
 };
 
-use super::{InEnvironment, InferOk, InferResult, InferenceContext, TypeError};
+use super::{InEnvironment, InferOk, InferenceContext, TypeError};
+
+pub(crate) type CoerceResult = Result<InferOk<(Vec<Adjustment>, Ty)>, TypeError>;
+
+/// Do not require any adjustments, i.e. coerce `x -> x`.
+fn identity(_: Ty) -> Vec<Adjustment> {
+    vec![]
+}
+
+fn simple(kind: Adjust) -> impl FnOnce(Ty) -> Vec<Adjustment> {
+    move |target| vec![Adjustment { kind, target }]
+}
+
+/// This always returns `Ok(...)`.
+fn success(
+    adj: Vec<Adjustment>,
+    target: Ty,
+    goals: Vec<InEnvironment<Goal<Interner>>>,
+) -> CoerceResult {
+    Ok(InferOk { goals, value: (adj, target) })
+}
 
 impl<'a> InferenceContext<'a> {
     /// Unify two types, but may coerce the first one to the second one
     /// using "implicit coercion rules" if needed.
-    pub(super) fn coerce(&mut self, from_ty: &Ty, to_ty: &Ty) -> bool {
+    pub(super) fn coerce(&mut self, from_ty: &Ty, to_ty: &Ty) -> CoerceResult {
         let from_ty = self.resolve_ty_shallow(from_ty);
         let to_ty = self.resolve_ty_shallow(to_ty);
         match self.coerce_inner(from_ty, &to_ty) {
-            Ok(result) => {
-                self.table.register_infer_ok(result);
-                true
+            Ok(InferOk { value, goals }) => {
+                self.table.register_infer_ok(InferOk { value: (), goals });
+                Ok(InferOk { value, goals: Vec::new() })
```

Comment:
this is kind of weird, since it's handling half of the result. can we record the adjustments here as well?